### DISCUSSION TITLE
How to show each page only once on search results?

### DISCUSSION THREAD

**Discussion by kaycebasques** [sphinx-doc/sphinx#12022] [Q&A]
If you search for `pw_unit_test` on https://pigweed.dev/search.html you get results like this:

![image](https://github.com/sphinx-doc/sphinx/assets/4713486/5c60e76f-8524-4b0c-94bb-5028accfe9b9)

This happens because we have structured our [changelog](https://pigweed.dev/changelog.html) like this:

* 20240226
  * ...
  * pw_unit_test
  * ...
* 20240212
  * ...
  * pw_unit_test
  * ...
 
`20240226` is an H2 containing the updates published on 2024 Feb 26. The `pw_unit_test` nested as a child is an H3 containing the updates for that module. Etc.

Is there a way to configure the local search engine to only show each page once?

**Comment by picnixz** [comment]
AFAIK, no. But there are a bunch of PRs trying to fix the search engine so maybe they can give you better results.

**Comment by jayaddison** [comment]
I think that this problem is a symptom of a (recently-resolved) bug https://github.com/sphinx-doc/sphinx/issues/11943 : in Sphinx versions up-to-and-including 7.2.6, identical contextual snippet appeared despite search results linking to multiple matches within a document.

**Reply by jayaddison** [reply]
See also bugreport #11961 about search result duplicates - with two possible pull requests currently available (#11942, #12047) that perform corresponding de-duplication.  Feedback on either of those would be much appreciated.

**Reply by jayaddison** [reply]
> I think that this problem is a symptom of a (recently-resolved) bug #11943 : in Sphinx versions up-to-and-including 7.2.6, identical contextual snippet appeared despite search results linking to multiple matches within a document.

After further investigation: resolving #11943 does **not** de-duplicate the results mentioned in this thread, because each of them is a subtitle/subheading in the document, and we currently (up-to-and-including v7.4.6) include each distinct subtitle/subheading result (both pull requests mentioned addressed de-duplication of main-document-title results only).

I think there are two problems here:

  * The results aren't useful when a user performs the search query.
  * Each affected result creates a separate HTTP GET request to retrieve the same page for the purposes of displaying a summary (wasteful).

**Reply by jayaddison** [reply]
Although it might seem that we should de-duplicate the HTTP GET (or, strictly speaking: browser `fetch` operations) requests without making any changes to the eventual display of results, I think there is a user experience consideration to bear in mind:

We add each result one-by-one to the HTML search result display output.  If we were to begin de-duplicating `fetch` requests without making any other changes, then the _first_ request to retrieve the `changelog` page here would be slow -- but then all of the following subsection headings would rapidly appear and fill the user's display, and _then_ slow fetch operations for other results would commence.

That might not be a great search experience, because each result could appear with a wildly differing delay, and whether a result was fast/slow to retrieve would seem erratic.

Perhaps we could group the results by page, or by the type of match -- and perhaps allow the content in the grouping to update (or not!) as subsequent results are received.
